README.md
# ASWSG - Another|Alexanders Static WebSite Generator

ASWSG allows you to generate Websites using Markup Syntax and HTML.

* *make friendly* - to be integrated in your workflow
* *adaptable syntax* - to match different markup dialects

It is build with the idea of classical UNIX tools to do one job.
ASWSG behaves just like a compiler parsing a Markup file and generating a new HTML output.
ASWSG will not generate a filesystem structure, or an HTML frame, but you can do so using includes and
use it with a build system like *make*.


HTML and Markup can be mixed, reusable includes may be used for headers, footers and other repeating code blocks.
Most benefits come from using dynamic variables, which are partly set dynamically by aswsg,
by the document or via the command line and the include system.
In example the current *date*, *time* and the actual *filename* is set dynamically, *author*, *creation date*
and others in the document, the *sitename* as a command line parameter.
But it is even possible to redefine markup symbols due to different markup dialects.
(i.E. use "!" instead/as well of "=" for headers.)
All variables can be used subsequent, also in included files. (i.E. using the article name in a header include.)

This tool will generate new HTML code, which -- of course -- may contain dynamic code.

Note: This is under development, but what's not.

## Why should I use it

You like to create websites, and you want to have the freedom to build and use html the way you like it 
and don't want 

## Features

### Released

* [X] Simple markup parsing
  * [x] line based
    * [X] headers
    * [X] paragraphs
    * [X] ordered and unordered lists
    * [X] cites
    * [X] raw lines
    * [X] tables
    * [X] comments
  * [X] inline based
    * [X] bold
    * [X] emphasised
    * [X] strike trough
    * [X] underline
    * [X] links
      * [X] link type ``[[name|link]]`` (without rel=external)
      * [X] link type ``[name](link)`` (for external links with rel="external")
         * support link relationship attribute "external", see https://www.w3schools.com/tags/att_a_rel.asp
      * [x] link type ``[[URL]]``  with link-index
      * [X] link type ``=> URL description`` - a gemini style link
    * [X] block based
      * [X] quotes
      * [X] preformatted (code)
      * [X] raw html
      * [x] comments
* [X] include files
  * [X] include list of files by pattern 
* [X] Setting and using of variables
  * [X] Setting variables (in this order, later overwrites former)
    1. static defaults
    2. dynamically like filename
    3. from command line parameter
    4. in parsed file
  * [x] Using variables everywhere, also in commands, but not in raw imported files
* [X] Commands
  * [X] dump-vars  (to log)
  * [X] message  (to log)
  * [X] include raw or crude files (with variable parsing and replacing)
  * [X] Execute an external command and insert its output
* [X] Conditional parsing based on variables
* [X] redefine markup tags
* [x] Log Filter
* [X] Anchor command (for headers)
* [X] Import CSV-file as table

## Usage

`aswsg [IN-FILE:]sourcefile [VARIABLE:value] > file.html`

Checkout the example.

## Control

### Messages

Which messages will be logged, respectively not logged, can be controlled through the
variable ASWSG-MESSAGE-FILTER.
Messages with severity in ASWSG-MESSAGE-FILTER will not be sent to *stderr*.


## Formatting

Description of the markup formatting.

You can use the mentioned variables to redefine the characters used for a formatting.

### Line level formatting

Used at begin the beginning line, using one of the characters.
Some characters are can be cascaded.

| Function                                                                                    | Default Char | Variable | Example                                                                  |
|---------------------------------------------------------------------------------------------|-----| -------- |--------------------------------------------------------------------------|
| Defining a variable                                                                         | ```@``` | ```ASWSG-DEFINE``` | ```@variablename:value```                                                |
| Include a file oder files by pattern                                                        | ```+``` | ```ASWSG-INCLUDE``` | ```+filename``` or ```+file_patern```                                    |
| Raw (html) line to be inserted                                                              | ```$``` | ```ASWSG-RAWLINE``` | ```$<article>```                                                         |
| Escape for a paragraph char                                                                 | ```\``` | ```ASWSG-ESCAPE``` | ```\* this is no bullet list```                                          |
| Paragraph                                                                                   | (none) | (none) | ```Any text not starting not with a line level special. Empty lines start a new paragraph.``` |
| Header. The number of header characters define the depth of the header.                     | ```=``` or ```!``` | ```ASWSG-HEADER``` | ```== header level 2```                                                  |
| Bullet list                                                                                 | ```*``` or ```-``` | ```ASWSG-LIST``` | ```* Bulltes and numbered Lists may be nested.```                        |
| Numbered list                                                                               | any off ```#0123456789``` | ```ASWSG-NUMERATION``` | ```2# a level 2 indented list element```                                 |
| Cite                                                                                        | ```>``` | ```ASWSG-CITE``` | ```> To be or not to be.```                                              |
| Single line command, optionally closed by an ")", should not be changed                     | ```(``` | ```ASWSG-COMMAND``` | ```(command parameter ...)```                                            |
| Defining a Table. The table character starts a new cell. One trailing pipe will be removed. | ```\|``` | ```ASWSG-TABLE```                                                        | ```\|a 2 cell\|table``` |
 | Gemini style link                                                                           | ```=>``` | ```ASWSG-GEMINI-LINK``` | ```=> http://haniarani.com/ Hania Rani```                                |

To continue a long line (i.e. a long header split over two lines) add an ```\``` to the end of the first line.


### Inline formatting

Used to format text within a line, using 2 or 3 strings for begin, end and middle when needed.

| Function | Default Char | Variables | Example |
| -------- | ------------ | -------- | ------- |
| Variable replacing |  ```{{```...```}}``` | ```ASWSG-VAR-1```...```ASWSG-VAR-2``` | ```{{```variablename```}}```  |
| Bold | ```*```...```*``` | ```ASWSG-BOLD-1```...```ASWSG-BOLD-2``` | ```*```bold text```*```  |
| Emphasised | ```//```...```//``` | ```ASWSG-EMP-1```...```ASWSG-EMP-2``` | ```//```italic text```//```  |
| Code | ``` `` ```...``` `` ``` | ```ASWSG-CODE-1```...```ASWSG-CODE-2``` | ``` `` ```text``` `` ```  |
| Strike through | ```~~```...```~~``` | ```ASWSG-STRIKE-1```...```ASWSG-STRIKE-2``` | ```~~```text```~~```  |
| Underline | ```__```...```__``` | ```ASWSG-UNDERL-1```...```ASWSG-UNDERL-2``` | ```__```text```__```  |
| Link-1 (internal) | ```[[```...```\|```...```]]``` | ```ASWSG-LINK-1-1```...```ASWSG-LINK-1-3```...```ASWSG-LINK-1-2``` | ```[[```text```\|```URL```]]```  |
| Link-2 (with rel=external) | ```[```...```](```...```)``` | ```ASWSG-LINK-2-1```...```ASWSG-LINK-2-3```...```ASWSG-LINK-2-2``` | ```[```text```](```URL```)```  |
| Link-3 (Link Index, to be inserted with commnd (LINK-INDEX)) | ```[[```...```]]``` | ```ASWSG-LINK-3-1```...```ASWSG-LINK-1-2``` | ```[[```URL```]]```  |

#### Glitches

There is a glitch when using the default markdown for emphasised ```//``` and a HTML link, 
which usually contains also two dashes, in one line. 
There is a small fix to avoid, that // will be used as the start of emphasised text, 
when there is a ```:``` in front of it, but that prevents any later rendering of emphasised in this line.

A good and easy solution in this case is to redefine emphasised. 
This could be done like this, to use two $-signs instead:

    @ASWSG-EMP-1:$$
    @ASWSG-EMP-2:$$

### Multi-Line/Block formatting

A line just containing at least three characters to enter a special block.
The block ends with the same characters in a line or a new block formatting.

| Function                        | Default Char | Variable            | Example    |
|---------------------------------|--------------|---------------------|------------|
| Citation                        | ```>```      | ```ASWSG-ML-CITE``` | ```>>>```  |
| Raw Lines                       | ```$```      | ```ASWSG-ML-RAW```  | ```$$$```  |
| Preformatted (code)             | ```%```      | ```ASWSG-ML-CODE``` | ```%%%```  |
| Horizontal line (just one line) | ```-```      | ```ASWSG-LINE```    | ```----``` |

tbi = to be implemented, does not exist jet.

## Commands

All commands start with the ```ASWSG-COMMAND``` character which default is ```(```.

| Function | Form |
| -------- | ------------ |
| Comment that will not be in the output HTML file | ```COMMENT any text``` |
| Dump variables to log | ```DUMP-VARS parameters ignored``` |
| Write a message to the log | ```MESSAGE any text``` |
| Set an link anchor | ```ANCHOR anchor-name``` |
| Insert link index for Link-3 ```[[<LINK>>]]``` | ```LINK-INDEX parameters ignored``` |
| Include file raw without variable substitution | ```INCLUDE-FILE-RAW filename``` |
| Include file crude without variable substitution | ```INCLUDE-FILE-CRUDE filename``` | 
| Include the output of a script/command (no variable substitution) | ```INCLUDE-SCRIPT programm parameters``` |
| Include CSV-file as table | ```INCLUDE-CSV filename.csv``` |
| Use the line formatter ```>``` as ```<blockquote>``` instead off ```<cite>```. Activate this with the parameter "TRUE", "T" or no parameter. Every other parameter will be interpreted as false. (Default is ```cite```.) | ```GT-AS-BLOCKQUOTE``` | 

## Conditional commands

If a condition is set the following lines are only parsed if the condition criteria are true. 
This is valid until a new condition is set, a ```COND-ELSE``` or ```COND-END``` clause changes this.
There is only one condition valid at a time. 
There is no nesting, a new ```COND-IF-...``` just changes the condition.

All conditions start with the ```ASWSG-COMMAND``` character which default is ```(```.

| Function                                      | Form                                           |
|-----------------------------------------------|------------------------------------------------|
| Check if Variable is set                      | ```COND-IF-SET variable-name```                |
| Check if Variable is not set                  | ```COND-IF-NOT-SET variable-name```            |
| Check if Variable has the value               | ```COND-IF-EQUAL variable-name value...```     |
| Check if Variable has not the value           | ```COND-IF-NOT-EQUAL variable-name value...``` |
| Condition else case (last condition inverted) | ```COND-ELSE```                                |
| End condition validation                      | ```COND-END```                                 |

## More Variables

### Control Variables

| Function                                                                                                                   | Name:Default                                         |
|----------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------|
| Number of header lines when parsing a table.                                                                               | ```ASWSG-TABLE-HEADERLINES:1```                      |
| Alignment of colons when parsing a table. L=left, C=center, R=right. If the value is to short, or unkown it defaults to L. | ```ASWSG-TABLE-ALIGNMENT:LL```                       |
| CSV field separator                                                                                                        | ```ASWSG-CSV-COMMA:;```                              | 
| CSV comment line                                                                                                           | ```ASWSG-CSV-COMMENT:#```                            | 
| Automatically generate anchors for headers, T = true, everything else is false                                             | ```ASWSG-AUTO-GENERATE-ANCHOR:T```                   |
| Reverse Order of included files, T = true, everything else is false                                                        | ```ASWSG-INCLUDE-REVERSE:F```                        |
| Date format                                                                                                                | ```DATEFORMAT:2006-01-02```                          |
| Time format                                                                                                                | ```TIMEFORMAT:15:04:05```                            |

releases.md
# Releases
## Version - Released - Features

## V0.12.0 - 2022-08-23
- Gemini style links
- Bug fix in CODE und CITE blocks, HTML tags wrong nesting fixed
##  V0.11.1 - 2022-05-25
- process conditions before block mode (fixes two bugs)
### V0.11 - 2022-05-25
- Include several files with the `+` line command using pattern 
### V0.10.4 - 2022-05-20
- Bug fix counting same characters at line beginning
### V0.10.3 - 2021-09-12
- Wrong usage text `...[IN-FILE=]...` replaced with `...[IN-FILE:]...` in README and help-function. 
### V0.10.2 - 2021-09-11
- fix problem when no filename to process is give, plus some safeguards and hints
### V0.10.1 - 2021-08-26
- fix problem with empty lines
### V0.10 - 2021-08-25
- Command to change use from "cite" to "blockquote" for ">"
- Ignore comment line (Default starting with ";") 
- block modes for quotes, comments, code and raw html
### V0.9 - 2021-07
- Include CSV files as tables
### V0.8 - 2021-01 
- Conditional parsing with (COND-... 
### V0.7 - 2020-10
- Tables 
- Variable with line numbers
### V0.5 - 2020-09 
- Include output of external command
- file includes
### V0.3 - 2020-07
- Indexed Links 
### V0.2 - 2020-04
- Many things
### V0.0 - 2015-11
- First commit


aswsg.go
// Another|Alexanders Static WebSite Generator (aka aswsg)
// Dynamically generate HTML pages from markdown
// Copyright © 2016-2021 Alexander Kulbartsch
// License: AGPL-3.0-or-later (GNU Affero General Public License 3 or later)

/*
   This file is part of "Another|Alexanders Static WebSite Generator",
   also called "aswsg".

   aswsg is free software: you can redistribute it and/or modify it under the
   terms of the GNU Affero General Public License as published by the Free
   Software Foundation, either version 3 of the License, or any later version.

   aswsg is distributed in the hope that it will be useful, but WITHOUT ANY
   WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
   FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for
   more details.

   You should have received a copy of the GNU Affero General Public License
   along with aswsg.  If not, see <http://www.gnu.org/licenses/>.
*/

// Basic function:
//
// Lines beginning with "@" (and no further white space)
// are interpreted as variables in the form "@var: value".
// White spaces after the double colon is optional and gets removed.
//
// Lines beginning with a "+" (and no further white space)
// are interpreted in the form "+filename".
// The named file will be included here.
//
// \ escapes the special line characters (and will be removed)
//
// A {{variable}} in the text will be replaced by the named variable

// TODO: use OUT-FILE (will be done with context type) - maybe

package main

import (
	"bufio"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

type siteContextType struct {
	vars               SimpleVars
	inStream           io.Reader // ???
	outStream          io.Writer // ???
	paragraphState     string
	lineNumber         int
	blockMode          string
	tableLine          int
	conditionFulfilled bool
}

var siteContext siteContextType

var paragraphTags = map[string]string{
	" ": "",
	"B": "li",
	"C": "cite", // can be changed to blockquote with command
	"D": "td",
	"L": "ul", // "ul style=\"list-style-type:circle\"",
	"N": "ol",
	"O": "pre",
	"P": "p",
	"R": "tl",
	"T": "table",
	"b": "b",
}

type indexedLinksType struct {
	index string
	link  string
}

var indexedLinks []indexedLinksType
var linkIndex int = 1

// message handling

// Message logs to stderr
func Message(filename string, line int, severity string, messagetext string) {
	if !strings.ContainsAny(severity, siteContext.vars.GetVal("ASWSG-MESSAGE-FILTER")) {
		var fn string
		var ln int
		if filename == "" {
			fn = siteContext.vars.GetVal("filename")
		} else {
			fn = filename
		}
		if line == 0 {
			ln = siteContext.lineNumber
		} else {
			ln = line
		}
		log.Println(fn, ":", ln, ":", severity, ":", messagetext)
	}
}

// main

// site context helper functions

func (c siteContextType) addStringToOutput(s string) (err error) {
	// ToDo implement
	return
}

func setDefaultSiteVars() {
	siteContext.vars = SimpleVars{
		"ASWSG-VERSION": "0.12.0",
		"ASWSG-AUTHOR":  "Alexander Kulbartsch",
		"ASWSG-LICENSE": "AGPL V3 or later",

		// control vars
		"ASWSG-MESSAGE-FILTER":       "Dd", // D = Debug
		"ASWSG-AUTO-GENERATE-ANCHOR": "T",  // T = true, everything else is false
		"ASWSG-TABLE-HEADERLINES":    "1",  // number of headers, when parsing a table.
		"ASWSG-TABLE-ALIGNMENT":      "LL", // L -> <th style="text-align:left">, C = center, R = right, other/default = left
		"ASWSG-CSV-COMMA":            ";",  // CSV field separator
		"ASWSG-CSV-COMMENT":          "#",  // CSV comment line
		"ASWSG-INCLUDE-REVERSE":      "F",  // reverse order of files to be included

		// inline formating, pairs end on -1 respective -2
		"ASWSG-VAR-1":    "{{", // special: variable to be replaced
		"ASWSG-VAR-2":    "}}",
		"ASWSG-LINK-1-1": "[[", // special: link internal
		"ASWSG-LINK-1-2": "]]",
		"ASWSG-LINK-1-3": "|",
		"ASWSG-LINK-2-1": "[", // special: link
		"ASWSG-LINK-2-2": ")",
		"ASWSG-LINK-2-3": "](",
		"ASWSG-LINK-3-1": "[[", // special: indexed-link
		"ASWSG-LINK-3-2": "]]",
		"ASWSG-BOLD-1":   "*", // inline: bold
		"ASWSG-BOLD-2":   "*",
		"ASWSG-EMP-1":    "//", // inline: emphasised
		"ASWSG-EMP-2":    "//",
		"ASWSG-CODE-1":   "``", // inline: code
		"ASWSG-CODE-2":   "``",
		"ASWSG-STRIKE-1": "~~", // inline: strike through
		"ASWSG-STRIKE-2": "~~",
		"ASWSG-UNDERL-1": "__", // inline: underline
		"ASWSG-UNDERL-2": "__",

		// line level formating (for paragraphs) at begin of line, using one of the characters
		"ASWSG-DEFINE":   "@",  // special: define var
		"ASWSG-INCLUDE":  "+",  // special: include parsed file
		"ASWSG-CONTINUE": "\\", // special: if at end of line, continue (join) with next line
		// "ASWSG-RAWFILE": "<",  // special: include raw file - will not be implemented this way, but as command. This special character will be used to identify raw HTML code. See ASWSG-RAWHMTL.
		"ASWSG-RAWHMTL": "<",  // TODO special: raw html line (this may have leading white spaces)
		"ASWSG-RAWLINE": "$",  // special: raw (html) line
		"ASWSG-ESCAPE":  "\\", // special: escape char for paragraph
		// ... paragraph: initial state: __ (empty)
		// ... paragraph: _P_aragraph
		"ASWSG-LIST":       "*-",          // paragraph: _L_ist and _B_ullets
		"ASWSG-CITE":       ">",           // paragraph: _C_ite
		"ASWSG-NUMERATION": "#0123456789", // paragraph: _N_umbered list and _B_ullets
		"ASWSG-COMMAND":    "(",           // single line command, optionally closed by an ")". Symbold should not be changed
		"ASWSG-TABLE":      "|",           // paragraph: _T_able and _R_ows and D_ata
		"ASWSG-HEADER":     "=!",          // one liner: header
		"ASWSG-COMMENT":    ";",           // comment line
		"ASWSG-NEWLINE":    "/",           // TODO: insert a <br /> tag

		"ASWSG-GEMINI-LINK":      "=>", // gemini-style link
		"ASWSG-GEMINI-LINK-SHOW": "F",  // show the link itself

		// Block level formatting: unique multi characters in one line alone, at least 3
		"ASWSG-LINE":       "-", // special: horizontal line
		"ASWSG-ML-CODE":    "%", // start/end block: code code
		"ASWSG-ML-CITE":    ">", // start/end block: cite mention
		"ASWSG-ML-CRUDE":   "$", // start/end block: raw line (for HTML code)
		"ASWSG-ML-COMMENT": ";", // start/end block: comment lines
	}
	_ = siteContext.vars.SetVar("TimeStampFormat", "2006-01-02 15:04:05 UTC+ 07:00")
	_ = siteContext.vars.SetVar("DateFormat", "2006-01-02")
	_ = siteContext.vars.SetVar("TimeFormat", "15:04:05")
	_ = siteContext.vars.SetVar("now", time.Now().Format(siteContext.vars.GetVal("TimeStampFormat")))
	_ = siteContext.vars.SetVar("today", time.Now().Format(siteContext.vars.GetVal("DateFormat")))
	_ = siteContext.vars.SetVar("time", time.Now().Format(siteContext.vars.GetVal("TimeFormat")))
}

func parseAndSetCommandLineVars() (exitAswsg bool) {
	destinationVar := "IN-FILE"
	for i := 1; i < len(os.Args); i++ {
		arg := os.Args[i]
		Message("$CMDLINEARG$", i, "D", arg)
		if arg[0] == '-' {
			showHelp()
			return true
		} else if strings.Index(arg, ":") >= 0 {
			if siteContext.vars.ParseAndSetVar(arg) != true {
				Message("$CMDLINEARG$", i, "w", "Can't parse variable: "+arg)
			}
		} else {
			if siteContext.vars.SetVar(destinationVar, arg) != true {
				Message("$CMDLINEARG$", i, "w", "Can't parse variable: "+arg)
			} else {
				if destinationVar == "IN-FILE" {
					destinationVar = "OUT-FILE"
				} else {
					Message("$CMDLINEARG$", i, "W", "To much non variable parameters (ignored): "+arg)
				}
			}
		}
	}
	return false
}

// vars

func parseAndSetVar(line string) (varParsed bool) {
	if strings.ContainsAny(line[0:1], siteContext.vars.GetVal("ASWSG-DEFINE")) {
		siteContext.vars.ParseAndSetVar(line[1:])
		return true
	}
	return false
}

func replaceInlineVars(line string) string {
	t1, t2, t3 := StringBracketsSplit(line, siteContext.vars.GetVal("ASWSG-VAR-1"), siteContext.vars.GetVal("ASWSG-VAR-2"), siteContext.vars.GetVal("ASWSG-ESCAPE"))
	if !siteContext.vars.ExistsVal(t2) {
		return line
	}
	return replaceInlineVars(t1 + siteContext.vars.GetVal(t2) + t3)
}

// line

// changeParagraphs returns the necessary HTML Tags to close the previous state and initiate the new one.
// if both states are the same refreshInner forces the inner tag to be closed and opened
func changeParagraphs(oldParagraphState string, newParagraphState string, refreshInner bool) (resultLines []string) {
	if oldParagraphState == newParagraphState && len(oldParagraphState) > 0 {
		if refreshInner {
			resultLines = append(resultLines, generateTag(Right(oldParagraphState, 1), false))
			resultLines = append(resultLines, generateTag(Right(newParagraphState, 1), true))
		}
		return
	}
	intermediateState := oldParagraphState
	Message("", 0, "D", "intermediateState: '"+intermediateState+"', newParagraphState: '"+newParagraphState+"'")
	// close previous paragraph state(s)
	for len(intermediateState) > 0 {
		iSLen := len(intermediateState)
		if iSLen <= len(newParagraphState) && intermediateState == newParagraphState[:iSLen] {
			break
		}
		x := len(intermediateState) - 1
		resultLines = append(resultLines, generateTag(Right(intermediateState, 1), false))
		intermediateState = intermediateState[:x]
	}
	// open new paragraph state(s)
	for len(intermediateState) < len(newParagraphState) /* && intermediateState != newParagraphState */ {
		x := len(intermediateState)
		addState := newParagraphState[x : x+1]
		resultLines = append(resultLines, generateTag(addState, true))
		intermediateState = intermediateState + addState
	}
	return
}

// parse paragraph line + parse inline
func parseCommonParagraphControls(line string, currentParagraphState string) (resultLines []string, resultingParagraphState string) {
	firstChar := line[0:1]
	resultingParagraphState = ""
	refreshInner := false
	surroundWith := ""
	controlChars := siteContext.vars.GetVal("ASWSG-LIST") + siteContext.vars.GetVal("ASWSG-CITE") + siteContext.vars.GetVal("ASWSG-NUMERATION")
	// Message("", 0, "D", "inline:" + line)
	// Message("", 0, "D", "  PS:" + currentParagraphState)

	// parse LIST, CITE and NUMERATION
	for _, r := range line {
		a := string(r)
		Message("", 0, "D", "  controlChar:"+a)
		if ContainsOnly(a, controlChars) {
			switch {
			case ContainsOnly(a, siteContext.vars.GetVal("ASWSG-LIST")):
				resultingParagraphState = resultingParagraphState + "L"
				surroundWith = "li"
			case ContainsOnly(a, siteContext.vars.GetVal("ASWSG-CITE")):
				resultingParagraphState = resultingParagraphState + "C"
			case ContainsOnly(a, siteContext.vars.GetVal("ASWSG-NUMERATION")):
				resultingParagraphState = resultingParagraphState + "N"
				surroundWith = "li"
			default:
				Message("", 0, "A", "should not happen (controlChar not found)")
				break
			}
			line = WhiteSpaceTrim(line[1:])
		} else {
			break
		}
		// Message("", 0, "D", "  resultState:"+ resultingParagraphState )
	}

	// parse Escape
	if firstChar == siteContext.vars.GetVal("ASWSG-ESCAPE") {
		line = line[1:]
	}

	// parse paragraph
	if len(resultingParagraphState) == 0 && len(line) != 0 {
		resultingParagraphState = "P"
	}

	var newLine string
	if surroundWith != "" {
		newLine = surroundWithHTMLTag(surroundWith, parseInLine(line))
	} else {
		newLine = parseInLine(line)
	}

	resultLines = append(changeParagraphs(currentParagraphState, resultingParagraphState, refreshInner), newLine)

	return
}

// parse one line
func parseLine(line string, paragraphState string) (resultLines []string, newParagraphState string) {

	newParagraphState = paragraphState
	lineLength := len(line)

	// replace inline vars
	line = replaceInlineVars(line)

	// parseCondition
	if lineLength > 0 && line[0:1] == siteContext.vars.GetVal("ASWSG-COMMAND") {
		if parseCondition(line[1:]) {
			return resultLines, newParagraphState
		}
	}

	// validateCondition (return if not fulfilled)
	if !siteContext.conditionFulfilled {
		return resultLines, newParagraphState
	}

	// block modes

	blockToggle := checkBlockModeToggle(line)

	// block mode comment - "ASWSG-ML-COMMENT": ";"
	checkBlock := siteContext.vars.GetVal("ASWSG-ML-COMMENT")
	if siteContext.blockMode == checkBlock { // in this block
		if blockToggle == siteContext.blockMode { // end of this block
			siteContext.blockMode = ""
		} else { // normal in this block
			// comment does nothing
		}
		return
	} else if siteContext.blockMode == "" && blockToggle == checkBlock { // new block start
		siteContext.blockMode = checkBlock
		return
	}

	// block mode code - "ASWSG-ML-CODE":    "%"
	checkBlock = siteContext.vars.GetVal("ASWSG-ML-CODE")
	if siteContext.blockMode == checkBlock { // in this block
		if blockToggle == checkBlock { // end of this block
			siteContext.blockMode = ""
			newParagraphState = ""
			resultLines = append(changeParagraphs(paragraphState, newParagraphState, false), "</pre>")
			return resultLines, newParagraphState
		} else { // normal in this block
			resultLines = append(resultLines, "<code>"+line+"</code>")
		}
		return
	} else if siteContext.blockMode == "" && blockToggle == checkBlock { // new block start
		siteContext.blockMode = checkBlock
		newParagraphState = ""
		resultLines = append(changeParagraphs(paragraphState, newParagraphState, false), "<pre>")
		return resultLines, newParagraphState
	}

	// block mode crude - "ASWSG-ML-CRUDE":     "$"
	checkBlock = siteContext.vars.GetVal("ASWSG-ML-CRUDE")
	if siteContext.blockMode == checkBlock { // in this block
		if blockToggle == checkBlock { // end of this block
			// resultLines = append(resultLines, "</code>")
			siteContext.blockMode = ""
		} else { // normal in this block
			resultLines = append(resultLines, line)
		}
		return
	} else if siteContext.blockMode == "" && blockToggle == checkBlock { // new block start
		siteContext.blockMode = checkBlock
		return
	}

	// block mode cite - "ASWSG-ML-CITE":    ">"
	checkBlock = siteContext.vars.GetVal("ASWSG-ML-CITE")
	if siteContext.blockMode == checkBlock { // in this block
		if blockToggle == checkBlock { // end of this block
			siteContext.blockMode = ""
			newParagraphState = ""
			resultLines = append(changeParagraphs(paragraphState, newParagraphState, false), "</blockquote>")
			return resultLines, newParagraphState
		} else { // normal in this block
			// continue normal processing of line
		}
	} else if siteContext.blockMode == "" && blockToggle == checkBlock { // new block start
		siteContext.blockMode = checkBlock
		newParagraphState = ""
		resultLines = append(changeParagraphs(paragraphState, newParagraphState, false), "<blockquote>")
		return resultLines, newParagraphState
	}

	// empty line
	if lineLength == 0 {
		newParagraphState = ""
		resultLines = changeParagraphs(paragraphState, newParagraphState, false)
		return resultLines, newParagraphState
	}

	// ignore ; comment line
	if line[0:1] == siteContext.vars.GetVal("ASWSG-COMMENT") {
		return
	}

	// parse and set vars
	if parseAndSetVar(line) == true {
		return resultLines, newParagraphState
	}

	// parse commands
	if line[0:1] == siteContext.vars.GetVal("ASWSG-COMMAND") {
		resultLines = append(resultLines, parseCommands(line[1:])...)
		return
	}

	// Entering Markup Mode

	// parse raw/crude (variables where already replaced)
	if line[0:1] == siteContext.vars.GetVal("ASWSG-RAWLINE") {
		resultLines = append(resultLines, line[1:])
		return
	}

	// process includes
	if strings.ContainsAny(line[0:1], siteContext.vars.GetVal("ASWSG-INCLUDE")) {
		tmpLine := siteContext.lineNumber
		tmpFilename := siteContext.vars.GetVal("filename")

		files, err := filepath.Glob(line[1:])
		if err != nil {
			Message("", 0, "E", err.Error())
		}
		if IsVarTrue("ASWSG-INCLUDE-REVERSE") {
			files = ReverseStringArray(files)
		}

		for _, file := range files {
			parsedLines, parsedParagraph, err := parseFile(file, newParagraphState)
			if err != nil {
				Message(line[1:], -1, "E", err.Error())
			}
			resultLines = append(resultLines, parsedLines...)
			newParagraphState = parsedParagraph
		}

		siteContext.lineNumber = tmpLine
		siteContext.vars.SetVar("filename", tmpFilename)
		return resultLines, newParagraphState
	}

	// parse Markup one liner

	// parse one liner: gemini-style link =>
	if len(line) > 2 && line[0:2] == siteContext.vars.GetVal("ASWSG-GEMINI-LINK") {
		newParagraphState = "L"
		link, description := parseGeminiLink(line[2:])
		attrib := HTMLAttrib{"href": link}
		attribGemlink := HTMLAttrib{"class": "aswsg-gemlink"}
		var ht string
		if IsVarTrue("ASWSG-GEMINI-LINK-SHOW") {
			ht = surroundWithHTMLTagWithAttributes("li",
				surroundWithHTMLTagWithAttributes("a", link+" - "+description, attrib),
				attribGemlink)
		} else {
			ht = surroundWithHTMLTagWithAttributes("li",
				surroundWithHTMLTagWithAttributes("a", description, attrib),
				attribGemlink)
		}
		resultLines = append(changeParagraphs(paragraphState, newParagraphState, false), ht)
		return resultLines, newParagraphState
	}

	// parse one liner: header
	if strings.ContainsAny(line[0:1], siteContext.vars.GetVal("ASWSG-HEADER")) && (len(line) > 3) {
		newParagraphState = ""
		// header parser
		fc, count, content := firstCharCountAndTrim(line)
		if !ContainsOnly(fc, siteContext.vars.GetVal("ASWSG-HEADER")) {
			Message("", 0, "A", "should not happen - expected ASWSG-HEADER character")
		}
		level := strconv.Itoa(count)
		// ToC ~~~
		anchor := ""
		if siteContext.vars.GetVal("ASWSG-AUTO-GENERATE-ANCHOR") == "T" {
			anchor = " id=\"" + ToValidHtmlAnchor(content) + "\""
			// toc_line := strings.Repeat(siteContext.vars.GetVal("ASWSG-LIST"), level) + " (" + content + ")[" + anchorText + "]"
			// TODO add toc_line anchor to list
		}
		//
		resultLines = append(changeParagraphs(paragraphState, newParagraphState, false), "<h"+level+anchor+">"+parseInLine(content)+"</h"+level+">")
		return resultLines, newParagraphState
	}

	// parse Table
	if strings.ContainsAny(line[0:1], siteContext.vars.GetVal("ASWSG-TABLE")) {
		if paragraphState != "T" {
			siteContext.tableLine = 0
			newParagraphState = "T"
		}
		tline := parseTableLine(WhiteSpaceTrim(line[1:]))
		resultLines = append(changeParagraphs(paragraphState, newParagraphState, false), tline)
		return resultLines, newParagraphState
	}

	// parse one liner: horizontal line
	if ContainsOnly(strings.TrimRight(line, " \t"), siteContext.vars.GetVal("ASWSG-LINE")) && len(strings.TrimRight(line, " \t")) >= 3 {
		newParagraphState = " "
		resultLines = append(changeParagraphs(paragraphState, newParagraphState, false), "<hr \\>")
		return resultLines, newParagraphState
	}

	resultLines, newParagraphState = parseCommonParagraphControls(line, paragraphState)

	return resultLines, newParagraphState
}

// core logic /////////////////////////

func parseFile(filename string, startParagraphState string) ([]string, string, error) {
	var lines []string

	paragraphState := startParagraphState
	if len(filename) == 0 {
		Message("", 0, "W", "Filename missing!")
		showHelp()
		return nil, paragraphState, nil
	}

	file, err := os.Open(filename)
	if err != nil {
		return nil, paragraphState, err
	}
	defer file.Close()

	// TODO: set vars
	//    - filename        ok
	//    - fqfn
	//    - basefn
	//    - file change date + time
	//    - file creation date + time
	fileStat, statError := file.Stat()
	if statError != nil {
		return nil, paragraphState, err
	}
	siteContext.vars.SetVar("filename", fileStat.Name())

	scanner := bufio.NewScanner(file)
	// TODO check for errors

	var result []string
	var continued string
	siteContext.lineNumber = 0

	// stream scanner
	for scanner.Scan() {
		// TODO check for errors ?

		var oneInputLine string

		oneInputLine = continued + scanner.Text() // TODO check for Error
		siteContext.lineNumber += 1
		siteContext.vars.SetVar("linenumber", strconv.Itoa(siteContext.lineNumber))

		if Right(oneInputLine, 1) == siteContext.vars.GetVal("ASWSG-CONTINUE") {
			continued = oneInputLine[:len(oneInputLine)-1]
			continue
		}
		continued = ""

		lines, paragraphState = parseLine(oneInputLine, paragraphState)
		result = append(result, lines...)

	}

	return result, paragraphState, scanner.Err()

}

func showHelp() {
	println("aswsg - Another Static Website Generator - Version " + siteContext.vars.GetVal("ASWSG-VERSION"))
	println("Usage: aswsg [IN-FILE:]sourcefile [VARIABLE:value] > file.html")
	println("© " + siteContext.vars.GetVal("ASWSG-AUTHOR"))
	println("Licensensed under " + siteContext.vars.GetVal("ASWSG-LICENSE"))
	println("Find more information in the repo: https://github.com/Kulbartsch/AStaticWebSiteGenerator")
}

func main() {

	var parsedText []string
	var err error

	setDefaultSiteVars()

	Message("", 0, "D", "---- ASWSG start ----")

	paragraphState := ""
	siteContext.paragraphState = ""
	siteContext.conditionFulfilled = true
	siteContext.blockMode = ""

	if parseAndSetCommandLineVars() {
		os.Exit(1)
	}

	parsedText, paragraphState, err = parseFile(siteContext.vars.GetVal("IN-FILE"), paragraphState)
	if err != nil {
		fmt.Println("Error:", err.Error())
	}

	Message("", 9999, "D", "---- Resulting paragraph style :"+paragraphState)

	// cleanup unclosed paragraphs
	parsedText = append(parsedText, changeParagraphs(paragraphState, "", false)...)

	// Output
	// TODO use out file / stream
	for _, l := range parsedText {
		fmt.Println(l)
	}

	Message("", 0, "D", "---- bye ----")

}

//EOF
